### PR TITLE
Update cni spec version to 0.3.1 to support multiple ips

### PR DESCRIPTION
**Pull request by johscheuer**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change
> /kind bug

/kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:

This PR changes the default CNI spec for kubenet to `0.3.1` and allow multiple IPs as return result which is important for the DualStack Feature.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
kubenet uses now per default the CNI spec 0.3.1 and support multiple IPs (DualStack Feature)
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by johscheuer** [comment]
Currently there are no test cases for the setup method (I also don't know if this can be achieved easily). Is this something we actually want?

File: pkg/kubelet/dockershim/network/kubenet/kubenet_linux.go [hunk 1]
```diff
@@ -168,7 +168,7 @@ func (plugin *kubenetNetworkPlugin) Init(host network.Host, hairpinMode kubeletc
 	if err != nil {
 		klog.Warningf("can't set sysctl %s: %v", sysctlBridgeCallIPTables, err)
 	}
-
+	// make use of cni.go -> getLoNetwork(binDirs []string) *cniNetwork  ???
```
Comment (dcbw) [inline pkg/kubelet/dockershim/network/kubenet/kubenet_linux.go]: This seems like a leftover comment?

File: pkg/kubelet/dockershim/network/kubenet/kubenet_linux.go [hunk 2]
```diff
@@ -348,22 +347,18 @@ func (plugin *kubenetNetworkPlugin) setup(namespace string, name string, id kube
...
-	// that we get multiple IP addresses in the returned Result structure
-	if res.IP4 != nil {
-		ipv4 = res.IP4.IP.IP.To4()
-	}
 
-	if res.IP6 != nil {
-		ipv6 = res.IP6.IP.IP
+	for _, ip := range res.IPs {
+		// Add the IP to tracked IPs
+		plugin.addPodIP(id, ip.Address.IP.String())
```
Comment (dcbw) [inline pkg/kubelet/dockershim/network/kubenet/kubenet_linux.go]: @johscheuer 'res.IPs' might include addresses from interfaces *other* than the container interface. For each IP in the list, look at `ip.Interface` to find the index into the res.Interfaces list. If the item in the res.Interfaces list has a `sandbox != ""` then it's the container interface and you can use the IP address.

```
	for _, ipc := range res.IPs {
		if ipc.Interface == nil {
			continue
		}
		intIdx := *ipc.Interface
		if intIdx < 0 || intIdx >= len(res.Interfaces) || res.Interfaces[intIdx].Sandbox == "" {
			continue
		}
		<the IP address is a container/sandbox IP so use it>
```

File: pkg/kubelet/dockershim/network/kubenet/kubenet_linux.go [hunk 3]
```diff
@@ -348,22 +347,18 @@ func (plugin *kubenetNetworkPlugin) setup(namespace string, name string, id kube
...
 	}
 
-	if ipv4 == nil && ipv6 == nil {
-		return fmt.Errorf("cni didn't report ipv4 ipv6")
+	if res.IPs == nil || len(res.IPs) == 0 {
```
Comment (aojea) [inline pkg/kubelet/dockershim/network/kubenet/kubenet_linux.go]: should this check go before the for loop?
 if there are no res.IPs we don't have to iterate over res.IPs

File: pkg/kubelet/dockershim/network/kubenet/kubenet_linux.go (hunk 1 above)
Comment (johscheuer) [reply pkg/kubelet/dockershim/network/kubenet/kubenet_linux.go]: Ups, right. Was a reminder for me what's next.

File: pkg/kubelet/dockershim/network/kubenet/kubenet_linux.go (hunk 2 above)
Comment (johscheuer) [reply pkg/kubelet/dockershim/network/kubenet/kubenet_linux.go]: I just saw that https://github.com/kubernetes/kubernetes/pull/69821/files#diff-9768e7cd7bcf561500b1048edb02db5aR353-R362 has a similar method/function. Should this method be used here (as soon as it is merged)? Also I'm wondering why the checks are slightly different ?

File: pkg/kubelet/dockershim/network/kubenet/kubenet_linux.go (hunk 3 above)
Comment (johscheuer) [reply pkg/kubelet/dockershim/network/kubenet/kubenet_linux.go]: Good catch ! Do we actually need the ` len(res.IPs) == 0` check ? Theoretically we should always get at least one result (loopback).

File: pkg/kubelet/dockershim/network/kubenet/kubenet_linux.go (hunk 2 above)
Comment (dcbw) [reply pkg/kubelet/dockershim/network/kubenet/kubenet_linux.go]: @johscheuer the logic is the same, though the code is slightly different. I think I prefer https://github.com/kubernetes/kubernetes/pull/69821/files#diff-9768e7cd7bcf561500b1048edb02db5aR353-R362 so just use that one.

File: pkg/kubelet/dockershim/network/kubenet/kubenet_linux.go (hunk 3 above)
Comment (dcbw) [reply pkg/kubelet/dockershim/network/kubenet/kubenet_linux.go]: @johscheuer you don't need the `nil` check; just `len(res.IPs) == 0` is fine.

File: pkg/kubelet/dockershim/network/kubenet/kubenet_linux.go (hunk 2 above)
Comment (AlexeyPerevalov) [reply pkg/kubelet/dockershim/network/kubenet/kubenet_linux.go]: @dcbw, my suggestion is to move that helper function to some common place.

File: pkg/kubelet/dockershim/network/kubenet/kubenet_linux.go (hunk 3 above)
Comment (johscheuer) [reply pkg/kubelet/dockershim/network/kubenet/kubenet_linux.go]: Changed it, thanks!

File: pkg/kubelet/dockershim/network/kubenet/kubenet_linux.go [hunk 4]
```diff
@@ -67,13 +67,13 @@ const (
 	zeroCIDRv4 = "0.0.0.0/0"
 
 	NET_CONFIG_TEMPLATE = `{
-  "cniVersion": "0.1.0",
+  "cniVersion": "0.3.1",
+  "supportedVersions": [ "0.3.0" ],
```
Comment (dcbw) [inline pkg/kubelet/dockershim/network/kubenet/kubenet_linux.go]: @johscheuer you shouldn't need supportedVersions here, that's actually what plugins themselves return (in this case 'bridge').

Comment (johscheuer) [reply pkg/kubelet/dockershim/network/kubenet/kubenet_linux.go]: Removed

**Comment by danwinship** [comment]
ping @dcbw

File: pkg/kubelet/dockershim/network/kubenet/kubenet_linux.go (hunk 1 above)
Comment (dcbw) [reply pkg/kubelet/dockershim/network/kubenet/kubenet_linux.go]: @johscheuer can you remove this comment?

Comment (johscheuer) [reply pkg/kubelet/dockershim/network/kubenet/kubenet_linux.go]: Updated, could you take another look?

File: pkg/kubelet/dockershim/network/kubenet/kubenet_linux.go (hunk 2 above)
Comment (dcbw) [reply pkg/kubelet/dockershim/network/kubenet/kubenet_linux.go]: @johscheuer still need to update the code above to handle https://github.com/kubernetes/kubernetes/pull/82844/files#r329772998

**Comment by dcbw** [comment]
@johscheuer PR still needs an update to ignore non-sandbox interfaces per https://github.com/kubernetes/kubernetes/pull/82844#discussion_r329772998

**Comment by johscheuer** [comment]
Good catch, I updated the PR with the referenced changes.

**Review by cmluciano — COMMENTED** [review]
Unfortunately there are not any unit tests for this setup function but based on the latest updates to the cni codebase, this change looks good to me.

File: pkg/kubelet/dockershim/network/kubenet/kubenet_linux.go (hunk 3 above)
Comment (AlexeyPerevalov) [reply pkg/kubelet/dockershim/network/kubenet/kubenet_linux.go]: check for res.IPs == nil is not necessary, cause len(nil) is 0. 
Got it, you put changes to separate commit "Correct check if CNI result has IPs", why not to amend a commit?

**Review by AlexeyPerevalov — CHANGES_REQUESTED** [review]
I would prefer to see amended commits (git commit -amend)

**Comment by johscheuer** [comment]
@AlexeyPerevalov if you want I can squash some of the commits (and change the commit message)

**Comment by AlexeyPerevalov** [comment]
> @AlexeyPerevalov if you want I can squash some of the commits (and change the commit message)

yes, it would be great

**Comment by johscheuer** [comment]
@cmluciano I added some unit tests for the specific part

**Review by AlexeyPerevalov — APPROVED** [review]
Thanks, Johannes now it's lgtm

**Comment by johscheuer** [comment]
@dcbw any news on this one?

**Comment by khenidak** [comment]
@johscheuer i see some tests are failing.. can i ask you to ensure that they are cleared?

**Comment by johscheuer** [comment]
@khenidak which tests are failing for you?

When I run the tests locally everything works:

```bash
=== RUN   TestGetPodNetworkStatus
--- PASS: TestGetPodNetworkStatus (0.00s)
=== RUN   TestTeardownCallsShaper
    TestTeardownCallsShaper: kubenet_linux_test.go:198: PASS:   DelNetwork(mock.AnythingOfTypeArgument,mock.AnythingOfTypeArgument,mock.AnythingOfTypeArgument)
--- PASS: TestTeardownCallsShaper (0.00s)
=== RUN   TestInit_MTU
--- PASS: TestInit_MTU (0.00s)
=== RUN   TestTearDownWithoutRuntime
    TestTearDownWithoutRuntime: kubenet_linux_test.go:318: PASS:        DelNetwork(mock.AnythingOfTypeArgument,mock.AnythingOfTypeArgument,mock.AnythingOfTypeArgument)
    TestTearDownWithoutRuntime: kubenet_linux_test.go:318: PASS:        DelNetwork(mock.AnythingOfTypeArgument,mock.AnythingOfTypeArgument,mock.AnythingOfTypeArgument)
    TestTearDownWithoutRuntime: kubenet_linux_test.go:318: PASS:        DelNetwork(mock.AnythingOfTypeArgument,mock.AnythingOfTypeArgument,mock.AnythingOfTypeArgument)
    TestTearDownWithoutRuntime: kubenet_linux_test.go:318: PASS:        DelNetwork(mock.AnythingOfTypeArgument,mock.AnythingOfTypeArgument,mock.AnythingOfTypeArgument)
--- PASS: TestTearDownWithoutRuntime (0.05s)
=== RUN   TestGetRoutesConfig
--- PASS: TestGetRoutesConfig (0.00s)
=== RUN   TestGetRangesConfig
--- PASS: TestGetRangesConfig (0.00s)
=== RUN   TestTrackPodIPs
--- PASS: TestTrackPodIPs (0.00s)
PASS
ok      k8s.io/kubernetes/pkg/kubelet/dockershim/network/kubenet        0.081s
```

**Comment by cmluciano** [comment]
ping @dcbw for review

**Comment by johscheuer** [comment]
If there is anything I can help with just ping me. One good thing about the "new" CNI version is that we can refactor some parts of kubenet to use the common CNI plugins with chaining (like stated in some comments).

File: pkg/kubelet/dockershim/network/kubenet/kubenet_linux.go [hunk 5]
```diff
@@ -329,12 +328,31 @@ func (plugin *kubenetNetworkPlugin) Capabilities() utilsets.Int {
...
+		intf := res.Interfaces[*ip.Interface]
+		if intf.Sandbox != "" && intf.Name == network.DefaultInterfaceName {
+			// Found our sandbox interface, use the IP
+			podGateways = append(podGateways, ip.Gateway)
+			podCIDRs = append(podCIDRs, net.IPNet{IP: ip.Address.IP.Mask(ip.Address.Mask), Mask: ip.Address.Mask})
+			plugin.addPodIP(id, ip.Address.IP.String())
+		}
+	}
+
+	return podGateways, podCIDRs
```
Comment (khenidak) [inline pkg/kubelet/dockershim/network/kubenet/kubenet_linux.go:351]: That will blank out PodIPs. Is this what we want?

Comment (johscheuer) [reply pkg/kubelet/dockershim/network/kubenet/kubenet_linux.go:351]: The `plugin.addPodIP` is called directly in the for loop instead afterwards so the behaviour should be the same?

**Comment by aojea** [comment]
does it makes sense to continue with this?
seems dockershim is moving out next release https://github.com/kubernetes/kubernetes/pull/94624

**Comment by aojea** [comment]
> does it makes sense to continue with this?
> seems dockershim is moving out next release #94624

@johscheuer should we close this?

**Pull request closed without merging**